lib/linter-registry.coffee

```diff
@@ -0,0 +1,68 @@
+{Emitter} = require('atom')
+validate = require('./validate')
+helpers = require('./helpers')
+
+class LinterRegistry
+  constructor: ->
+    @linters = []
+    @locks =
+      Regular: new WeakSet
+      Fly: new WeakSet
+    @emitter = new Emitter
+
+  getLinters: ->
+    return @linters.slice() # Clone the array
+
+  hasLinter: (linter) ->
+    @linters.indexOf(linter) isnt -1
+
+  addLinter: (linter) ->
+    try
+      validate.linter(linter)
+      @linters.push(linter)
+    catch e then helpers.error(e)
+
+  deleteLinter: (linter) ->
+    return unless @hasLinter(linter)
+    @linters.splice(@linters.indexOf(linter), 1)
+
+  lint: ({onChange, editorLinter}) ->
+    editor = editorLinter.editor
+    lockKey = if onChange then 'Fly' else 'Regular'
+    return if onChange and not atom.config.get('linter.lintOnFly')
+    return unless editor is atom.workspace.getActiveTextEditor()
+    return unless editor.getPath()
+    return if @locks[lockKey].has(editorLinter)
+
+    @locks[lockKey].add(editorLinter)
+    scopes = editor.scopeDescriptorForBufferPosition(editor.getCursorBufferPosition()).scopes
+    scopes.push('*') # To allow global linters
+
+    # Confusing code ahead, proceed with caution :P
+    return @linters.reduce((promise, linter) =>
+      if helpers.shouldTriggerLinter(linter, true, onChange, scopes)
+        return promise.then(-> linter.lint(editor)).then((vals) =>
+          if vals then @emitter.emit('did-update-messages', {linter, messages: vals, editor})
+        ).catch (e) -> helpers.error(e)
+      else return promise
+    , Promise.resolve()).then( =>
+      Promises = []
+      @linters.forEach (linter) =>
+        return unless helpers.shouldTriggerLinter(linter, false, onChange, scopes)
+        try result = linter.lint(editor) catch e then return helpers.error(e)
```

Comment:
I recommend replacing this very awkward code with a form of what was here originally.

```CoffeeScript
        return new Promise((resolve) =>
          resolve(linter.lint(@editor, Helpers))
        ).then((results) => )
        .catch((e) -> helpers.error(e) # <- this line is new
```

1. the code above is easier to read
2. it doesn't have any of this checking to see if it's a promise and wrapping it nonsense.